File: apps/api/app/models/board.py
from __future__ import annotations

from datetime import date
from typing import Any

from sqlalchemy import JSON, Date, Float, ForeignKey, Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDMixin


class BoardColumn(UUIDMixin, TimestampMixin, Base):
    """A stage in the content pipeline Kanban (e.g. Ideas, Editing, Scheduled)."""

    __tablename__ = "board_column"

    workspace_id: Mapped[str] = mapped_column(
        ForeignKey("workspace.id", ondelete="CASCADE"), index=True, nullable=False
    )
    name: Mapped[str] = mapped_column(nullable=False)
    color: Mapped[str] = mapped_column(default="slate", nullable=False)
    position: Mapped[int] = mapped_column(Integer, default=0, nullable=False)


class BoardCard(UUIDMixin, TimestampMixin, Base):
    """A content idea/draft card living in a board column."""

    __tablename__ = "board_card"

    workspace_id: Mapped[str] = mapped_column(
        ForeignKey("workspace.id", ondelete="CASCADE"), index=True, nullable=False
    )
    column_id: Mapped[str] = mapped_column(
        ForeignKey("board_column.id", ondelete="CASCADE"), index=True, nullable=False
    )
    title: Mapped[str] = mapped_column(nullable=False)
    notes: Mapped[str | None] = mapped_column(nullable=True)
    # Float ordering so cards can be re-sequenced without renumbering siblings.
    position: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)

    # --- Content production fields (Notion-style card detail) ---
    emoji: Mapped[str | None] = mapped_column(nullable=True)
    # Free-text status label (e.g. "Idea", "Ready to record", "Filming"). Column-level
    # status comes from which column the card lives in; this is finer-grained.
    status: Mapped[str | None] = mapped_column(nullable=True)
    # Distribution targets (e.g. ["Instagram", "YouTube Shorts"]).
    platforms: Mapped[list[Any]] = mapped_column(JSON, default=list, nullable=False)
    publish_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    # The opening line of the script (most important sentence). Separate from notes.
    hook: Mapped[str | None] = mapped_column(nullable=True)
    # Visual concept description for the reel.
    visual_hook: Mapped[str | None] = mapped_column(nullable=True)
    caption: Mapped[str | None] = mapped_column(nullable=True)
    hashtags: Mapped[list[Any]] = mapped_column(JSON, default=list, nullable=False)
    reference_url: Mapped[str | None] = mapped_column(nullable=True)
    raw_footage_url: Mapped[str | None] = mapped_column(nullable=True)
    cover_image_url: Mapped[str | None] = mapped_column(nullable=True)

File: apps/web/lib/server/board.ts
import { videoAnalysisOut } from "./videoAnalysis";

// Shared Content Board card serialization (Prisma row → client shape). Kept in
// one place so the board list, the card PATCH, and the analyze/script routes all
// emit the same fields - including the new tags, scripted_at and the watched
// reference-reel analysis.

interface VaRow {
  status: string;
  summary: string | null;
  analysis: unknown;
  transcript: string | null;
  analyzedAt?: Date | null;
  error?: string | null;
}

export interface CardRow {
  id: string;
  columnId: string;
  title: string;
  notes: string | null;
  position: number;
  emoji: string | null;
  status: string | null;
  platforms: unknown;
  publishDate: string | null;
  hook: string | null;
  visualHook: string | null;
  caption: string | null;
  hashtags: unknown;
  referenceUrl: string | null;
  rawFootageUrl: string | null;
  coverImageUrl: string | null;
  scriptedAt: Date | null;
  tags: unknown;
  videoAnalysis?: VaRow | null;
}

export function serializeCard(c: CardRow) {
  return {
    id: c.id,
    column_id: c.columnId,
    title: c.title,
    notes: c.notes,
    position: c.position,
    emoji: c.emoji,
    status: c.status,
    platforms: (c.platforms as string[]) ?? [],
    publish_date: c.publishDate,
    hook: c.hook,
    visual_hook: c.visualHook,
    caption: c.caption,
    hashtags: (c.hashtags as string[]) ?? [],
    reference_url: c.referenceUrl,
    raw_footage_url: c.rawFootageUrl,
    cover_image_url: c.coverImageUrl,
    scripted_at: c.scriptedAt ? c.scriptedAt.toISOString() : null,
    tags: (c.tags as string[]) ?? [],
    video_analysis: videoAnalysisOut(c.videoAnalysis ?? null),
  };
}
